examples/gpio/config.c

```diff
@@ -0,0 +1,248 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/* This examples demonstrates the use of GPIO available on some TPM modules */
+
+#include <wolftpm/tpm2_wrap.h>
+
+#if defined(WOLFTPM_ST33) || defined(WOLFTPM_AUTODETECT)
+
+#include <examples/gpio/gpio.h>
+#include <examples/tpm_io.h>
+#include <examples/tpm_test.h>
+
+#include <stdio.h>
+#include <stdlib.h> /* atoi */
+
+/******************************************************************************/
+/* --- BEGIN TPM2.0 GPIO Configuration example  -- */
+/******************************************************************************/
+
+static void usage(void)
+{
+    printf("Expected usage:\n");
+    printf("./examples/gpio/gpio [num] [mode]\n");
+    printf("* num is a GPIO number between 0-3 (default %d)\n", TPM_GPIO_A);
+    printf("* mode is a number selecting the GPIO mode between 0-5 (default %d):\n", TPM_GPIO_MODE_STANDARD);
+    printf("\t0. standard - reset to the GPIO's default mode\n");
+    printf("\t1. floating - input in floating configuration.\n");
+    printf("\t2. pullup   - input with pull up enabled\n");
+    printf("\t3. pulldown - input with pull down enabled\n");
+    printf("\t4. opendrain - output in open drain configuration\n");
+    printf("\t5. pushpull  - output in push pull configuration\n");
+    printf("Demo usage, without parameters, configures GPIO%d as pushpull output.\n", TPM_GPIO_A);
+}
+
+int TPM2_GPIO_Config_Example(void* userCtx, int argc, char *argv[])
+{
+    int rc = -1;
+    int gpioNum = 0;
+    WOLFTPM2_DEV dev;
+    WOLFTPM2_CAPS caps;
+    WOLFTPM2_NV nv;
+    WOLFTPM2_HANDLE parent;
+    TPMI_GPIO_MODE gpioMode = TPM_GPIO_MODE_STANDARD;
+    TPM_HANDLE nvIndex = TPM_NV_GPIO_SPACE;
+    GpioConfig_In gpio;
+    SetCommandSet_In setCmdSet;
+    word32 nvAttributes;
+    int gpioInput = 0;
+    BYTE dummy = 0;
+
+   if (argc >= 2) {
+        if (XSTRNCMP(argv[1], "-?", 2) == 0 ||
+            XSTRNCMP(argv[1], "-h", 2) == 0 ||
+            XSTRNCMP(argv[1], "--help", 6) == 0) {
+            usage();
+            return 0;
+        }
+        if (argc == 3) {
+            gpioMode = atoi(argv[2]);
+            if (gpioMode > 5) {
+                printf("GPIO mode is out of range (0-5)\n");
+                usage();
+                goto exit_badargs;
+            }
+            /* Check if mode is for GPIO Input to perform extra config step */
+            if (gpioMode > 0 && gpioMode < 4) {
+                gpioInput = 1;
+            }
+            gpioMode += TPM_GPIO_MODE_STANDARD;
+            /* Preparing to process next argument */
+            argc--;
+        }
+        if (argc == 2) {
+            gpioNum = atoi(argv[1]);
+            if (gpioNum < 0 || gpioNum > 3) {
```

Comment:
Can we make the GPIO pin number min/max macros? These 0-3 are hard coded several places and should have option to change at build-time. Actually looks like min is always `TPM_GPIO_A` and max just needs added in tpm2.h.